### ISSUE TITLE
Google Nest Thermostat HomeKit Temperature Conversion Unit of Measurement Issue

### ISSUE THREAD

**Issue by mkormendy** [robertklep/name.klep.homekitty#105]
I have a Google Nest Thermostat added to my Homey via the Google Nest application. The Google Nest device is set to Fahrenheit, and so is my HomeKit too.
I have HomeKitty exposing the Nest Thermostat to HomeKit, but I'm seeing a bit of a conversion issue.
For example, the current temperature is 76º F but for some reason, that's being read as a Celsius number of 76º C to which is being displayed on my HomeKit Thermostat current temperature as 1169º F.

Wondering if there's a fix for this.

See below:

![49304A51-8D26-4D80-9655-401A8F043913_1_102_o](https://github.com/user-attachments/assets/09a31d5a-1c28-47b6-92b9-046681ba3419)

![D20CEE3D-2CBA-4015-AE10-43B2768AF5A6_1_102_o](https://github.com/user-attachments/assets/11f74551-0168-4e5e-a115-73813caa1155)

![FFB9EBCB-5FAF-45E4-B9D2-CCA9FCD4BE27_1_102_o](https://github.com/user-attachments/assets/ad8f9598-8b3e-41d9-a6e4-b432193d8f51)

![4615E77B-4917-4424-8CC0-4C52328CE83B_1_102_o](https://github.com/user-attachments/assets/bad8fef0-e8c3-47d6-9aa9-3b8bc202e194)

![27A5A64E-8C06-4D21-B601-867C00EE454C_1_102_o](https://github.com/user-attachments/assets/d481d56a-13e0-4052-ae7d-8917a640ca7d)

**Comment by robertklep** [comment]
My guess is that this is an issue with the Nest app.

Can you find the device here: https://tools.developer.homey.app/tools/devices

And then check if the temperature capabilities (`measure_temperature` and/or `target_temperature`) are set to the temperature in F?

**Comment by mkormendy** [comment]
Yes both of them are set to 76°

**Comment by robertklep** [comment]
That's not right, Homey internally should always store temperatures in Celcius, that's how the capability is defined:
<img width="356" alt="image" src="https://github.com/user-attachments/assets/4f3de539-1a60-43ed-850a-8f5bd327e7c2">

Which means that the Nest app is importing the temperature but doesn't check its scale and (apparently) assumes it's Celcius. Since HomeKitty also assumes Celcius (because the capability definition says so), iOS thinks that the supplied temperature is in C and converts it to F, causing an incorrect temperature to be shown.

I can't fix this in HomeKitty.

**Comment by mkormendy** [comment]
Let me reach out to the homey nest app developers and reference this issue as an example. What's interesting is that the Homey mobile app sees the thermostat temperature for cooling as Fahrenheit as well as the local weather temperature too at the top of the app. I'm not sure if this is automatically adjusted based on iOS system settings, but I can't seem to find a preference setting for this in the mobile app either.

![image](https://github.com/user-attachments/assets/55c1e6f2-c83b-46fb-bc85-81c7c021b015)

![image](https://github.com/user-attachments/assets/f23888e4-2418-451a-9a83-ad9c55f0c485)

**Comment by robertklep** [comment]
I'm fairly sure that the app uses your Homey's location to determine which temperature scale to use.

However, it's a bit unexpected that the Homey app shows the correct temperature, because I would think that it also, just like HomeKitty, also assumes that Celcius is used internally. Which would mean it would also convert the 76° to 169°.

What I'm just noticing though is that only the measured temperature is set incorrectly in your initial screenshot, the target temperature looks okay:

<img width="399" alt="image" src="https://github.com/user-attachments/assets/f5143711-fcc7-42bf-8bbb-a9e01d626f7b">

That's weird.

**Comment by mkormendy** [comment]
Yeah that is also very odd.
I was on my phone earlier when I was looking at the homey devices properties page and only saw the numbers for the thermostat .. but here's a snap of the full details:

<img width="1501" alt="Screenshot 2024-08-29 at 10 38 08 AM" src="https://github.com/user-attachments/assets/880c7517-8d38-47d8-b417-8b0ce9c81b29">

**Comment by robertklep** [comment]
I took a look at the code of the Nest app, and it looks like it uses the ambient temperature in Celcius, as provided by the Nest API, to set `measure_temperature`, which means the app is doing the right thing.

However, that doesn't explain why the developer tools is showing the temperature in Fahrenheit. It might be possible that the developer tools converts these values based on the user's locale, but I doubt it.

It could also be that the Nest API is providing an incorrect value, but I have no way of checking this.

I've asked one of the Athom developers on Slack if they can provide some clarification on what I'm seeing.

**Comment by robertklep** [comment]
Can you check and see if this version fixes your problem?

https://homey.app/a/name.klep.homekitty/test/

**Comment by mkormendy** [comment]
Installed the update, and now the temperature units are the same in the Home app, albeit both now converted improperly somewhere.
And I also noticed the following:
- the modes option in Home app doesn't quite reflect the modes in Homey or the thermostat itself
- the hardware display feature in the Home app thermostat settings doesn't really do anything when I switch it

![F7F6D821-CC87-412E-9852-8A0ADA6BA228_4_5005_c](https://github.com/user-attachments/assets/60f4d3a6-3c65-4961-ae0c-8a5da69eba90)
![A5D25B65-4D25-404D-A697-3E5EF02FB187_4_5005_c](https://github.com/user-attachments/assets/e222b233-bf88-4f0c-8871-eb669473329c)
![D8B921A6-4758-4563-9F78-504E88098DD9_4_5005_c](https://github.com/user-attachments/assets/247e5aaf-dcab-4097-babf-59db64522252)
![A21B64EB-A676-45E7-B19A-CD0A3EE27A40_4_5005_c](https://github.com/user-attachments/assets/3c08190a-1f97-47e4-8c7f-a2612df3121a)
![0034BB2C-CBCF-4343-877C-43FA3051A40A_4_5005_c](https://github.com/user-attachments/assets/e95f28b9-43ee-4b91-9b7a-86bd917c4de7)
![15D63882-1117-40BC-AADA-D5AA1B570125_4_5005_c](https://github.com/user-attachments/assets/cb53d555-75c2-4628-889c-41623887f6f0)

**Comment by robertklep** [comment]
There are various issues here:
* There seems to be a bug in Homey's core that causes the incorrect temperatures. The issue is that although the unit is set to Celcius, Homey's core provides the temperature values in Fahrenheit nonetheless. I've asked Athom developers to confirm that this is a bug, because it makes it impossible for me to determine if a temperature should be passed as C or F
* Because of the temperature issue, the Home UI doesn't work correctly either.
* The Nest app uses a custom thermostat mode capability instead of Homey's built-in one. HomeKitty doesn't support custom capabilities.

**Comment by mkormendy** [comment]
The first two points, I understand, however the last point, I'm not sure what to do about. Should I tell the Nest app devs that their custom capability makes it difficult for other apps to work with their app in Homey?

**Comment by robertklep** [comment]
I assume that they don't use the standard `thermostat_mode` capability for a reason, but you can always ask them.

There's not much you, or I, can do about this to get the device working properly in HomeKit though, at least not without resorting to something like a virtual/"advanced" device that provides a mapping between the custom capability to a regular capability: you would basically create a virtual thermostat device and set up flows that will update it when you change the actual thermostat, and also the other way around, when you change the virtual thermostat (from iOS) you update the actual thermostat.

**Comment by mkormendy** [comment]
Interesting .. I'm not exactly sure how to contact the developers of this to speak with them about it.

**Comment by robertklep** [comment]
Contact info can usually be found on the app page in the app store.

**Comment by mkormendy** [comment]
Okay, I have found it, it is Athom BV that made it .. I submitted a request and received the automated response below:

> Hi Michael,
> 
> Thanks for reaching out! Your request (124695) has been received and is being reviewed by our support staff.
> 
> Our estimated response time is 3 business days.

**Comment by robertklep** [comment]
@mkormendy I'm still waiting for something at Athom to [respond to my findings](https://github.com/athombv/homey-web-api-issues/issues/58), but it's becoming more and more difficult for developers like me to get any response (many issues get reported but go unanswered, as Athom seems to prioritise new features).

**Comment by mkormendy** [comment]
Do I need to also chime in on slack and or the github post?

**Comment by robertklep** [comment]
@mkormendy thanks for the offer, but not necessary 😊 Did you ever get a response from Athom support about this?

**Comment by mkormendy** [comment]
I did not. 😞

**Comment by robertklep** [comment]
At least you're getting a nice dashboard 😛

**Comment by suqbus** [comment]
I'm not sure if it's allowed to write here but on my side my Nest thermostat is always off in Homekit while it is in heating mode. If I put it on heating mode from Homekit it always returns to off mode on HomeKit, while it is in heating mode on Homey and in the Nest app.

**Comment by robertklep** [comment]
@suqbus this is probably related to the Nest app not using the standard `thermostat_mode` capability but a custom one, and HomeKitty doesn't support custom capabilities ☹️

**Comment by suqbus** [comment]
Last winter, however, everything worked perfectly. I turned off the thermostat for the summer and it was when I turned it on this week that I realized this new problem. Maybe the Nest app made changes to this recently?

**Comment by robertklep** [comment]
@suqbus that's very possible, I don't see any relevant changes to the thermostat code in HomeKitty that could explain this.

**Comment by mkormendy** [comment]
Looking at the Homey Pro updates changelog v12.2.0, they had some temperature fixes for an issue that could cause temperature conversion from Celsius to Fahrenheit to return an invalid value.
Does that help correct our problem?

**Comment by robertklep** [comment]
@mkormendy I got a diagnostics report from another user about this issue, and they are running v12.2.1. I'll try and see if I can force my Homey to use F as a unit so I can at least reproduce the issue.

I also bumped up [my issue about this](https://github.com/athombv/homey-web-api-issues/issues/58) once again. It's only been almost two months... 🤷🏼‍♂️

**Comment by ParkFans** [comment]
That was me, I also found this same issue with their builtin Alexa Integration. I did submit a support ticket with them for that, I'll let you know if I hear anything.

**Comment by GavinMcNabb** [comment]
That was me, I also found this same issue with their builtin Alexa Integration. I did submit a support ticket with them for that, I'll let you know if I hear anything.

**Comment by robertklep** [comment]
I think this might be another issue then.

Looking back at your comment [here](https://github.com/robertklep/name.klep.homekitty/issues/105#issuecomment-2317083560), I was wrong in assuming that the value shown on the "devices" page should have been in Celcius. I put my Homey in Fahrenheit mode and it was showing the value in F on that page, even though internally it was returning the proper value (in Celcius), so the devices page isn't showing the actual capability value, but a localised one.

I just realise that I only tested with the target temperature though, and looking back at this conversation the issue was with the _measured_ temperature. So I'll do some more testing.

**Comment by mkormendy** [comment]
Can you provide a switch in the app that a user can select that will perform a re-calculation to ºF for thermostats?

**Comment by robertklep** [comment]
I can't do that, because I don't know what's causing the issue and if it applies to _all_ thermostats or just a specific brand.

**Comment by mkormendy** [comment]
Let the user determine that, they can turn it on if they are seeing the issue.

**Comment by mkormendy** [comment]
In the Homey mobile app, I am seeing the temperature correctly.
![image](https://github.com/user-attachments/assets/ad395b5a-1bcc-433f-90a2-8753e6968c63)

**Comment by robertklep** [comment]
I can't reproduce the problem, both the Homey app and the Home app on iOS are showing the correct temperatures. I don't have an actual Nest thermostat that I can use to test things with.

**Comment by robertklep** [comment]
What we can try is that you make some changes to your thermostat's target temperature, then create a diagnostics report for Homekitty from the Homey app. Homekitty logs a fair amount of data, perhaps there's a clue in there.

**Comment by mkormendy** [comment]
Ah shucks, when you said you didn't have a nest thermostat to test with, I then decided to move over to a homebridge sdm implementation which has been working perfectly even with the mode and fan switch.

**Comment by robertklep** [comment]
@mkormendy no problem, I can imagine that at some point you just want a working device 😅

**Comment by robertklep** [comment]
@Tartsee we weren't able to fix the problem, I know there's a conversion issue somewhere but I don't know exactly where. 

You can try what I proposed in [this comment](https://github.com/robertklep/name.klep.homekitty/issues/105#issuecomment-2500196808).

**Comment by Tartsee** [comment]
Thank you for your reply. Forgive me, but I’m not sure I understand the part about “make some changes to target temp.”  Do you mean change the desired temp using HomeKit? As much as I really wanted this thermostat in HK, I completely understand if you would rather not bother with this issue.  I so appreciate everything else your app allows me to do.  Please let me know if you would like to continue.  Thank you again.  On Feb 21, 2025, at 8:55 PM, Robert Klep ***@***.***> wrote:﻿
  @Tartsee we weren't able to fix the problem, I know there's a conversion issue somewhere but I don't know exactly where.
You can try what I proposed in this comment.—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>



robertklep left a comment (robertklep/name.klep.homekitty#105)
@Tartsee we weren't able to fix the problem, I know there's a conversion issue somewhere but I don't know exactly where.
You can try what I proposed in this comment.

**Comment by robertklep** [comment]
@Tartsee yes, change the desired temperature. That will cause the HomeKitty app to log some information, which is not visible for users but is available in diagnostics reports.

**Comment by Tartsee** [comment]
Okay, I changed the temp inside HomeKit and sent the diagnostics.  I was provided with the following code:

1ac9445f-c480-4eef-9eca-d226fc3b5729

I do feel like this is a problem with the official Nest app because when I tried to use a virtual thermostat (not fully understanding how to do that), it showed the same current temp, which was incorrect.

Thank you, again, for looking into this.

________________________________

**Comment by robertklep** [comment]
@Tartsee it looks like Homey is using Fahrenheit internally, which it shouldn't be doing. I'm fairly sure now that this is an issue with the Nest app.

This is what was being logged:
```
[HomeKitty DEA6@Thermostat@Thermostat@Target Temperature] characteristic was supplied illegal value: number 68 exceeded maximum of 38
```

Which means that HomeKitty received the value "68" from Homey, which is the temperature in F, but it should be in C (because that's how Homey stores temperatures internally).

**Comment by robertklep** [comment]
@Tartsee I would also advise you to install the current test version of the app, you can install it from here: https://homey.app/a/name.klep.homekitty/test/

It's more up-to-date and has some temperature-related changes that _may_ solve your problem (although I don't think they will).

**Comment by Tartsee** [comment]
I suspect you are correct about the problem being on the Nest/Homey end.  I will notify the Homey developers, if the Homekitty test app does not solve the problem.  If it does solve the problem, I will notify you. Thank you very much for your time & assistance.Sent from my Verizon iPhoneOn Feb 22, 2025, at 10:12 PM, Robert Klep ***@***.***> wrote:﻿
  @Tartsee I would also advise you to install the current test version of the app, you can install it from here: https://homey.app/a/name.klep.homekitty/test/
It's more up-to-date and has some temperature-related changes that may solve your problem (although I don't think they will).—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>



robertklep left a comment (robertklep/name.klep.homekitty#105)
@Tartsee I would also advise you to install the current test version of the app, you can install it from here: https://homey.app/a/name.klep.homekitty/test/
It's more up-to-date and has some temperature-related changes that may solve your problem (although I don't think they will).

**Comment by Tartsee** [comment]
I tried out the test version of your app. It did not help with the Nest.  My actual physical target setting was 68, showed up in HomeKit as 154. The current temperature showed on the physical device as 70 and showed in HomeKit as 158.Again, thank you for your efforts. I will reach out to Homey and see if I get anywhere with them. Have a good night/day…Sent from my Verizon iPhoneOn Feb 22, 2025, at 10:12 PM, Robert Klep ***@***.***> wrote:﻿
  @Tartsee I would also advise you to install the current test version of the app, you can install it from here: https://homey.app/a/name.klep.homekitty/test/
It's more up-to-date and has some temperature-related changes that may solve your problem (although I don't think they will).—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>



robertklep left a comment (robertklep/name.klep.homekitty#105)
@Tartsee I would also advise you to install the current test version of the app, you can install it from here: https://homey.app/a/name.klep.homekitty/test/
It's more up-to-date and has some temperature-related changes that may solve your problem (although I don't think they will).
